tendermint/src/lite/verifier.rs

```diff
@@ -55,84 +87,33 @@ where
 }
 
 /// Verify the commit is valid from the given validators for the header.
-pub fn verify<H, V, C>(header: H, commit: C, validators: V) -> Result<(), Error>
+fn verify_header_and_commit<H, V, C>(header: &H, commit: &C, validators: &V) -> Result<(), Error>
 where
     H: Header,
     V: ValidatorSet,
     C: Commit,
 {
-    if let Err(e) = validate_vals_and_commit(header, &commit, &validators) {
+    if let Err(e) = validate_vals_and_commit(header, commit, validators) {
         return Err(e);
     }
 
     // ensure that +2/3 validators signed correctly
-    verify_commit_full(&validators, commit)
-}
-
-/// Verify the commit is trusted according to the last validators and is valid
-/// from the current validators for the header.
-pub fn verify_trusting<H, V, C>(
-    header: H,
-    commit: C,
-    last_validators: V,
-    validators: V,
-) -> Result<(), Error>
-where
-    H: Header,
-    V: ValidatorSetLookup,
-    C: Commit,
-{
-    // NOTE it might be more prudent to do the cheap validations first
-    // before we even call verify_commit_trusting, but not doing that
-    // makes the code cleaner and allows us to just call verify directly.
-
-    // ensure that +1/3 of last trusted validators signed correctly
-    if let Err(e) = verify_commit_trusting(&last_validators, &commit) {
-        return Err(e);
-    }
-
-    // perform same verification as in sequential case
-    verify(header, commit, validators)
+    verify_commit_full(validators, commit)
 }
 
 /// Verify that +2/3 of the correct validator set signed this commit.
 /// NOTE: these validators are expected to be the correct validators for the commit.
-fn verify_commit_full<V, C>(vals: &V, commit: C) -> Result<(), Error>
+fn verify_commit_full<V, C>(vals: &V, commit: &C) -> Result<(), Error>
 where
     V: ValidatorSet,
     C: Commit,
 {
     let total_power = vals.total_power();
-    let mut signed_power: u64 = 0;
-
-    let vals_vec = vals.into_vec();
-    let commit_vec = commit.into_vec();
-
-    if vals_vec.len() != commit_vec.len() {
+    if vals.len() != commit.votes_len() {
         return Err(Error::InvalidCommitLength);
     }
 
-    // The vals and commit have a 1-to-1 correspondence.
-    // This means we don't need the validator IDs or to do any lookup,
-    // we can just zip the iterators.
-    let vals_iter = vals_vec.iter();
-    let commit_iter = commit_vec.iter();
-    for (val, vote_opt) in vals_iter.zip(commit_iter) {
-        // skip absent and nil votes
-        // NOTE: do we want to check the validity of votes
-        // for nil ?
-        let vote = match vote_opt {
-            Some(v) => v,
-            None => continue,
-        };
-
-        // check vote is valid from validator
-        let sign_bytes = vote.sign_bytes();
-        if !val.verify_signature(&sign_bytes, vote.signature()) {
-            return Err(Error::InvalidSignature);
-        }
-        signed_power += val.power();
-    }
+    let signed_power = commit.voting_power_in(vals)?;
```

Comment:
@ancazamfir @ebuchman Does this work for you? I think it's closer to what is in the spec and should also be less confusing. It's used here and below in `verify_commit_trusting` (we do lookups in both cases now).

referencing related comments: https://github.com/interchainio/tendermint-rs/pull/36#discussion_r352581861

Reply:
Yes, it looks better imo.